### PR TITLE
Create Docker images for master components 

### PR DESCRIPTION
**Pull request by ArtfulCoder**
This is a step towards containerization of the various master components that run directly on host today.

### CODE REVIEW

**Comment by bprashanth** [comment]
assigning to Dawn for want of a better assignee, git blame wasn't too helpful

File: build/common.sh [hunk 1]
```diff
@@ -569,6 +578,30 @@ function kube::release::package_server_tarballs() {
...
+    for binary_name in "${KUBE_DOCKER_WRAPPED_BINARIES[@]}"; do
+      echo "+++ Building docker image: ${binary_name}";
+      local docker_file_path="$1/Dockerfile";
+      local binary_file_path="$1/${binary_name}";
+      if [ -f ${docker_file_path} ]; then
+        rm ${docker_file_path};
+      fi;
+      printf " FROM scratch \n ADD ${binary_name} /${binary_name} \n ENTRYPOINT [ \"/${binary_name}\" ]\n" >> ${docker_file_path};
+      local md5_sum=$(md5sum ${binary_file_path} | awk '{print $1}')
+      local docker_image_tag=kubernetes/$binary_name:$md5_sum
```
Comment (dchen1107) [inline build/common.sh]: s/kubernetes/gcr.io/google_containers?

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -28,6 +28,7 @@ readonly KUBE_SERVER_TARGETS=(
   cmd/kubernetes
   plugin/cmd/kube-scheduler
 )
+
```
Comment (dchen1107) [inline hack/lib/golang.sh]: Remove this extra line

Comment (ArtfulCoder) [inline hack/lib/golang.sh]: the extra line helps readability. Do you want me to create a separate PR to add a new line ?

File: build/common.sh (hunk 1 above)
Comment (ArtfulCoder) [inline build/common.sh]: Could you elaborate what you are recommending ? I didnt completely follow..

**Comment by ArtfulCoder** [comment]
@dchen1107  thanks for reviewing the PR! I added some comments..

File: build/common.sh (hunk 1 above)
Comment (ArtfulCoder) [inline build/common.sh]: sure.

File: build/common.sh [hunk 3]
```diff
@@ -569,6 +578,30 @@ function kube::release::package_server_tarballs() {
   done
 }
 
+# This will take binaries that run on master and creates Docker images
+# that wrap the binary in them. (One docker image per binary)
+function kube::release::create_docker_images_for_master() {
```
Comment (vmarmol) [inline build/common.sh]: nit: create_docker_images_for_wrapped_binaries() (since they don't necessarily have to be master, but are for now)

File: cluster/saltbase/install.sh [hunk 4]
```diff
@@ -24,6 +24,12 @@ set -o pipefail
 SALT_ROOT=$(dirname "${BASH_SOURCE}")
 readonly SALT_ROOT
 
+readonly DOCKER_IMAGES_FOR_MASTER=(
```
Comment (vmarmol) [inline cluster/saltbase/install.sh]: you use master here but docker_wrapped above

File: build/common.sh (hunk 1 above)
Comment (dchen1107) [inline build/common.sh]: I mean replace kubernetes with gcr.io/google_containers, so that the image is going to push to gcr.io, instead of docker hub.

Comment (ArtfulCoder) [inline build/common.sh]: We dont push at all to docker hub or gcr.io

File: cluster/saltbase/install.sh (hunk 4 above)
Comment (ArtfulCoder) [inline cluster/saltbase/install.sh]: yes, install.sh does not include any other scripts. so I cant easily share the array. I have another copy here.

File: hack/lib/golang.sh (hunk 2 above)
Comment (ArtfulCoder) [inline hack/lib/golang.sh]: Done

File: build/common.sh (hunk 3 above)
Comment (ArtfulCoder) [inline build/common.sh]: Done

File: build/common.sh (hunk 1 above)
Comment (dchen1107) [inline build/common.sh]: Yes, you are not pushing in this PR, but kubernetes here represents repository which should be gcr.io.

Comment (ArtfulCoder) [inline build/common.sh]: done

**Comment by dchen1107** [comment]
LGTM.

File: build/common.sh [hunk 5]
```diff
@@ -569,6 +578,30 @@ function kube::release::package_server_tarballs() {
...
+      echo "+++ Building docker image: ${binary_name}";
+      local docker_file_path="$1/Dockerfile";
+      local binary_file_path="$1/${binary_name}";
+      if [ -f ${docker_file_path} ]; then
+        rm ${docker_file_path};
+      fi;
+      printf " FROM scratch \n ADD ${binary_name} /${binary_name} \n ENTRYPOINT [ \"/${binary_name}\" ]\n" >> ${docker_file_path};
+      local md5_sum=$(md5sum ${binary_file_path} | awk '{print $1}')
+      local docker_image_tag=gcr.io/google_containers/$binary_name:$md5_sum
+      docker build -t "${docker_image_tag}" ${1};
```
Comment (derekwaynecarr) [inline build/common.sh]: this doesn't work unless the machine you are provisioning on has docker installed by default - i don't think this is the case for all salt-based providers.  this also doesn't make sense if we have salt scripts that are later responsible for installing docker as part of salt provisioning.

**Comment by derekwaynecarr** [comment]
So this PR broke the Vagrant cluster... are we requiring docker to be installed before doing a cluster/kube-up now?

**Comment by ArtfulCoder** [comment]
We do need docker to be installed to build docker images which then go into the server tarballs.
The salt-scripts install docker on the cluster, not on the machine which is building the tarballs.

Optionally, I could consider doing this as part of the deployment process, once salt-scripts have installed docker.

**Comment by derekwaynecarr** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/saltbase/salt/top.sls#L55

Looks to me that each kubernetes-master uses salt to install docker on the master , which is where this is erroring for me.

**Comment by ArtfulCoder** [comment]
Ah, yes
I will do the fix.

**Comment by zmerlynn** [comment]
@ArtfulCoder: I think one easy thing to do is just defer the "docker load" to a later Salt step. You can populate the pillar data as you're doing today, and just load after docker is ensured.

**Comment by mingfang** [comment]
FYI I have both the master and node dockerize here https://github.com/mingfang/docker-kubernetes-master and here https://github.com/mingfang/docker-kubernetes-node in case you're interested.
